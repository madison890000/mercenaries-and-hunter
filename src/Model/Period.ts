import Company from './Company';
import DataModel from './types';
import Base from './Base';

interface IPeriod {
    start: Date;
    end?: Date;
    keywords: DataModel.Keyword[];
    company: Company;
    jobPositionLevel: DataModel.JobPositionLevel;
    jobPosition: DataModel.JobPosition;
    achievements: DataModel.Achievement[];
    jobSummaries: string[];
    solutionsOfHowToImplement: DataModel.SolutionsOfHowToImplement[];
}

export default class Period extends Base {
    public start: Date;
    public end?: Date;
    public keywords: DataModel.Keyword[];
    public achievements: DataModel.Achievement[];
    public jobSummaries: string[];
    public company: Company;
    public jobPositionLevel: DataModel.JobPositionLevel;
    public jobPosition: DataModel.JobPosition;
    public solutionsOfHowToImplement: DataModel.SolutionsOfHowToImplement[];

    constructor({ solutionsOfHowToImplement, company, jobSummaries, jobPosition, jobPositionLevel, achievements, keywords, start, end }: IPeriod) {
        super();
        this.company = company;
        this.start = start;
        this.end = end;
        this.achievements = achievements;
        this.jobSummaries = jobSummaries;
        this.keywords = keywords;
        this.jobPositionLevel = jobPositionLevel;
        this.jobPosition = jobPosition;
        this.solutionsOfHowToImplement = solutionsOfHowToImplement;
    }
}
