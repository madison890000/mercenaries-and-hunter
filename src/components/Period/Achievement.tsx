import Tag from '../Tag';
import React from 'react';

interface AchievementProps {
    title: string;
    categories: string[];
}

const Achievement = ({ title, categories }: AchievementProps) => (
    <li>
        {title}
        <span>
            {categories?.map(c => (
                <Tag>{c}</Tag>
            ))}
        </span>
    </li>
);

export default Achievement;
