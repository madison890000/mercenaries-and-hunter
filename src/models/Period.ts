import Company from './Company';
import DataModel from './types';
import Base from './Base';
import StringWithID from './StringWithID';
import { v4 } from 'uuid';

interface IPeriod {
    start: Date;
    end?: Date;
    keywords: string[];
    company: Company;
    jobPositionLevel: DataModel.JobPositionLevel;
    jobPosition: DataModel.JobPosition;
    achievements: DataModel.IAchievement[];
    jobSummaries: string[];
    solutionsOfHowToImplement: DataModel.SolutionsOfHowToImplement[];
}

export default class Period extends Base {
    public start: Date;
    public end?: Date;
    public keywords: StringWithID[];
    public achievements: DataModel.Achievement[];
    public jobSummaries: StringWithID[];
    public company: Company;
    /**
     * @deprecated please use Job.level
     */
    public jobPositionLevel: DataModel.JobPositionLevel;
    /**
     * @deprecated please use Job.position
     */
    public jobPosition: DataModel.JobPosition;
    public job: DataModel.Job;
    public solutionsOfHowToImplement: DataModel.SolutionsOfHowToImplement[];

    constructor({
        solutionsOfHowToImplement,
        company,
        jobSummaries,
        jobPosition,
        jobPositionLevel,
        achievements,
        keywords,
        start,
        end
    }: IPeriod) {
        super();
        this.company = company;
        this.start = start;
        this.end = end;
        this.achievements = achievements.map(achievement => ({
            id: v4(),
            text: achievement.text,
            categories: achievement.categories.map(e => new StringWithID(e))
        }));
        this.jobSummaries = jobSummaries.map(e => new StringWithID(e));
        this.keywords = keywords.map(e => new StringWithID(e));
        this.jobPositionLevel = jobPositionLevel;
        this.jobPosition = jobPosition;
        this.job = {
            position: jobPosition,
            level: jobPositionLevel
        };
        this.solutionsOfHowToImplement = solutionsOfHowToImplement;
    }
}
