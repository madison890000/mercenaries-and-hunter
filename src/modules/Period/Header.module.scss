.header {
    font-weight: bold;
    display: flex;
    align-items: center;

    & > span {
        margin-right: var(--inline-block-margin);
    }

    @media screen and (max-width: 414px) {
        width: 300px;

        & > span {
            display: block;
        }
    }
}

.job {
    & > span {
        margin-right: var(--inline-block-margin);
    }
}

.companyName {
    font-weight: normal;
    color: var(--period-sub-title-color);
    display: flex;
    align-items: center;
}

.companyIndustry {
    font-weight: normal;
    color: var(--period-sub-title-color);
    font-size: 12px;
    margin: 6px;
}
