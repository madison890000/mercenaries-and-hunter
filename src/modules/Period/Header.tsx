import Tag from '../../components/Tag';
import React from 'react';
import DataModel from '../../models/types';
import styles from './Header.module.scss';
import capitalize from '../../utils/capitalize';
import Divider from '../../components/Divider';

interface HeaderProps {
    keywords: string[];
    jobPositionLevel: DataModel.JobPositionLevel;
    jobPosition: DataModel.JobPosition;
    companyName: string;
    companyIndustry: string;
    companyType: DataModel.CompanyType;
}

const Header = ({keywords, companyName, companyIndustry, jobPosition, jobPositionLevel}: HeaderProps) => (
    <div className={styles.header}>
        <div>
            <div className={styles.job}>
                {jobPositionLevel && <span>{capitalize(jobPositionLevel.toLowerCase())}</span>}
                <span>{jobPosition}</span>
            </div>
            <div className={styles.companyName}>
                <div>{companyName}</div>
                {/*<div className={styles.companyIndustry}>({capitalize(companyIndustry)})</div>*/}
            </div>
        </div>
        <Divider variant="v"/>
        <div style={{
            fontSize: 24
        }}>
            {keywords?.map(keyword => (
                <Tag type="filled" key={keyword}>
                    {keyword.toString()}
                </Tag>
            ))}
        </div>
    </div>
);

export default Header;
