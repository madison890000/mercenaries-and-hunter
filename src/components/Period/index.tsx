import React from 'react';
import styles from './index.module.scss';
import DataModel from '../../models/types';
import Times from '../Times';
import Achievement from './Achievement';
import TimeLineItem from '../TimeLines/TimeLineItem';
import { defineMessages, useIntl } from 'react-intl';
import Header from './Header';
import StringWithID from '../../models/StringWithID';
import capitalize from '../../utils/capitalize';

interface PeriodProps {
    start: Date;
    periodColor: string;
    end?: Date;
    keywords: StringWithID[];
    companyName: string;
    companyType: DataModel.CompanyType;
    jobPositionLevel: DataModel.JobPositionLevel;
    jobPosition: DataModel.JobPosition;
    achievements: DataModel.Achievement[];
    jobSummaries: StringWithID[];
}

const messages = defineMessages({
    jobSummaries: {
        id: 'component.period.jobSummaries',
        defaultMessage: 'Job summaries'
    },
    achievements: {
        id: 'component.period.achievements',
        defaultMessage: 'Achievements'
    }
});
const Period = ({
    start,
    periodColor,
    end,
    companyName,
    companyType,
    keywords,
    jobPositionLevel,
    jobPosition,
    achievements,
    jobSummaries
}: PeriodProps) => {
    const intl = useIntl();
    return (
        <div className={styles.periodContainer}>
            <div className={styles.period}>
                <div>
                    <Header
                        jobPosition={jobPosition}
                        companyName={companyName}
                        companyType={companyType}
                        keywords={keywords}
                        jobPositionLevel={jobPositionLevel}
                    />
                </div>
                <div className={styles.timeline}>
                    <TimeLineItem start={start} end={end} periodColor={periodColor} />
                    <Times start={start} end={end} />
                </div>
            </div>

            <div>
                <h5>{intl.formatMessage(messages.jobSummaries)}:</h5>
                <ul>
                    {jobSummaries?.map(summary => (
                        <li key={summary.id}>{capitalize(summary.toString())}</li>
                    ))}
                </ul>
            </div>
            <div>
                <h5>{intl.formatMessage(messages.achievements)}:</h5>
                <ul className={styles.achievements}>
                    {achievements?.map(achievement => (
                        <Achievement
                            key={achievement.id}
                            title={achievement?.text}
                            categories={achievement?.categories}
                        />
                    ))}
                </ul>
            </div>
        </div>
    );
};

export default Period;
