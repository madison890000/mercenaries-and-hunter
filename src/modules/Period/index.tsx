import React from 'react';
import {defineMessages, useIntl} from 'react-intl';

import styles from './index.module.scss';
import Achievement from '../../components/Achievement/Achievement';
import Header from './Header';

import capitalize from '../../utils/capitalize';
import pipe from '../../utils/pipe';
import {addPeriodSuffix} from '../../utils/suffix';

import Times from '../../components/Times';
import TimeLineItem from '../../components/TimeLines/TimeLineItem';
import Project from '../../components/Project';

import DataModel from '../../models/types';
import ProjectModel from '../../models/Project';

interface PeriodProps {
    start: Date;
    periodColor: string;
    end?: Date;
    keywords: string[];
    descriptions: string[];
    companyName: string;
    companyIndustry: string;
    companyType: DataModel.CompanyType;
    jobPositionLevel: DataModel.JobPositionLevel;
    jobPosition: DataModel.JobPosition;
    achievements: DataModel.IAchievement[];
    jobSummaries: string[];
    projects?: Omit<ProjectModel, 'id'>[];
}

const messages = defineMessages({
    jobSummaries: {
        id: 'component.period.jobSummaries',
        defaultMessage: 'Job summaries'
    },
    achievements: {
        id: 'component.period.achievements',
        defaultMessage: 'Achievements'
    }
});
const Period = ({
                    start,
                    periodColor,
                    end,
                    companyName,
                    companyType,
                    companyIndustry,
                    keywords,
                    descriptions,
                    jobPositionLevel,
                    jobPosition,
                    achievements,
                    jobSummaries,
                    projects
                }: PeriodProps) => {
    const intl = useIntl();
    return (
        <>
            <div className={styles.periodContainer}>
                <div className={styles.timeline}>
                    <TimeLineItem start={start} end={end} periodColor={periodColor}/>
                </div>
                <div className={styles.content}>
                    <div className={styles.period}>
                        <div>
                            <Header
                                jobPosition={jobPosition}
                                companyName={companyName}
                                companyType={companyType}
                                companyIndustry={companyIndustry}
                                keywords={keywords}
                                jobPositionLevel={jobPositionLevel}
                            />
                        </div>
                        <div className={styles.time}>
                            <Times start={start} end={end}/>
                        </div>
                    </div>
                    <div>
                        {descriptions?.map(description => (
                            <div className={styles.descriptionItem} key={description}>
                                {capitalize(description.toString())}
                            </div>
                        ))}
                    </div>
                    <div>
                        <ul>
                            {jobSummaries?.map(summary => (
                                <li
                                    key={summary}
                                    style={{
                                        marginRight: '2%',
                                        fontSize: 'var(--base-font-size-large-2)',
                                    }}
                                >
                                    {pipe<string>(capitalize, addPeriodSuffix)(summary.toString())}
                                </li>
                            ))}
                        </ul>
                    </div>
                    <div>
                        {achievements?.length > 0 && <div style={{
                            fontWeight: "bold",
                            fontSize: 'var(--base-font-size-middle)'
                        }}>{intl.formatMessage(messages.achievements)}:</div>}
                        <ul className={styles.achievements}>
                            {achievements?.map(achievement => (
                                <Achievement
                                    key={achievement.text}
                                    title={achievement?.text}
                                    categories={achievement?.categories}
                                />
                            ))}
                        </ul>
                    </div>
                    <div className={styles.projects}>
                        {projects?.map(project => (
                            <Project
                                projectName={project.name}
                                start={project.start}
                                end={project.end}
                                keywords={project.keywords}
                                descriptions={project.descriptions}
                                challengeAndSolutions={project.challengeAndSolutions}
                                achievements={project.achievements}
                            />
                        ))}
                    </div>
                </div>
            </div>
        </>
    );
};

export default Period;
