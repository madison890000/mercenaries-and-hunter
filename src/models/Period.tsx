import React from "react";
import Company from './Company';
import Base from './Base';
import EditText from './EditText';
import Times from './Times';
import Project from './Project';
import Description from "./Description";
import Achievement from "./Achievement";
import styled from "styled-components";
import {Divider, TimeLineItem} from "./components";
import {IAchievement, JobPosition} from "./types";
import Keywords from "./Keywords";
import ArrayData from "./ArrayData";

interface IPeriod {
    start: string;
    end?: string;
    keywords: string[];
    company: Company;
    periodColor?: string;
    jobPosition: JobPosition;
    achievements: IAchievement[];
    jobSummaries: string[];
    projects?: Project[];
    descriptions: string;
}

const Content = styled.div`
  padding: 10px;
  margin: 10px 0;
  border-radius: var(--base-border-radius);
  width: 100%;
`
const PeriodHeader = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-between;
`
const JobTitle = styled.div`
  font-size: var(--base-font-size-large);
`
const Time = styled.div`
  display: flex;
  justify-content: flex-end;
  align-items: center;
`

class Period extends Base {
    public keywords: Keywords;
    public descriptions: Description;
    public achievements: ArrayData<Achievement>;
    public jobSummaries: ArrayData<Description>;
    public company: Company;
    public job: EditText;
    public projects: ArrayData<Project>;
    public periodColor?: string;
    private times: Times;

    constructor({
                    company,
                    jobSummaries,
                    jobPosition,
                    achievements,
                    keywords,
                    descriptions,
                    start,
                    end,
                    periodColor,
                    projects
                }: IPeriod) {
        super();
        this.company = company;
        this.times = new Times(start, end).setParent(this);
        this.achievements = new ArrayData<Achievement>(
            achievements?.map(a => new Achievement(a?.text, a?.categories).setParent(this)),
            Achievement
        ).setParent(this);
        this.jobSummaries = new ArrayData<Description>(
            jobSummaries?.map(d => new Description(d, 'input').setParent(this)),
            Description,
        ).setParent(this);
        this.keywords = new Keywords(keywords).setParent(this);
        this.descriptions = new Description(descriptions, 'textarea', '职位简介').setParent(this)
        this.job = new EditText(jobPosition, 'input', '职位').setParent(this);
        this.projects = new ArrayData<Project>(projects ?? [], Project).setParent(this);
        this.periodColor = periodColor;
    }

    View = () => {
        return (
            <>
                <div style={{display: 'flex'}}>
                    <div style={{marginTop: 30}}>
                        <TimeLineItem start={this.times.start} end={this.times.end} periodColor={'black'}/>
                    </div>
                    <Content>
                        <PeriodHeader>
                            <div style={{display: "flex", alignItems: 'center'}}>
                                <JobTitle>
                                    <div>
                                        <span> <this.job.Show/></span>
                                    </div>
                                    <this.company.Show/>
                                </JobTitle>
                                <Divider variant="v"/>
                                <div>
                                    <this.keywords.Show/>
                                </div>
                            </div>
                            <Time>
                                <this.times.Show/>
                            </Time>
                        </PeriodHeader>
                        <div>
                            <this.descriptions.Show/>
                        </div>
                        <div>
                            <ul>
                                {this.jobSummaries?.data?.map(summary => <li>
                                    <summary.Show/>
                                </li>)}
                            </ul>
                        </div>
                        <div>
                            {this.achievements?.data?.length > 0 && <div>Achievements:</div>}
                            <ul style={{
                                fontSize: 'var(--base-font-size-middle)'
                            }}>
                                <this.achievements.Show/>
                            </ul>
                        </div>
                        <div>
                            <this.projects.Show/>
                        </div>
                    </Content>
                </div>
            </>
        )
    }
}

export default Period
