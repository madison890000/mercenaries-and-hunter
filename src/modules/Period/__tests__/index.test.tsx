import React from 'react';
import Period from '../index';
import renderer from 'react-test-renderer';
import DataModel from '../../../models/types';

test('renders Period', () => {
    const testCompanyName = 'test company name';
    const start = new Date('2012-01-01');
    const end = new Date('2013-01-01');
    const keywords = ['React'];
    const jobPositionLevel = DataModel.JobPositionLevel.Senior;
    const jobPosition = 'Software Engineer';
    const achievements = [] as DataModel.IAchievement[];
    const jobSummaries = [] as string[];
    const PeriodDom = renderer
        .create(
            <Period
                start={start}
                end={end}
                companyType={DataModel.CompanyType.Startup}
                keywords={keywords}
                jobPositionLevel={jobPositionLevel}
                jobPosition={jobPosition}
                achievements={achievements}
                jobSummaries={jobSummaries}
                companyName={testCompanyName}
                periodColor={'red'}
                companyIndustry={'tele'}
                descriptions={[]}/>
        )
        .toJSON();
    expect(PeriodDom).toMatchSnapshot();
});
