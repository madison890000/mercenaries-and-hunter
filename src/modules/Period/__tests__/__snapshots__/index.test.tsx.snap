// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Period 1`] = `
<div
  className="periodContainer"
>
  <div
    className="timeline"
  >
    <svg
      height={143}
      width={20}
    >
      <g>
        <rect
          fill="red"
          height="10"
          width={10}
          x={0}
          y={0}
        />
        <rect
          fill="red"
          height="10"
          width={10}
          x={0}
          y={11}
        />
        <rect
          fill="red"
          height="10"
          width={10}
          x={0}
          y={22}
        />
        <rect
          fill="red"
          height="10"
          width={10}
          x={0}
          y={33}
        />
        <rect
          fill="red"
          height="10"
          width={10}
          x={0}
          y={44}
        />
        <rect
          fill="red"
          height="10"
          width={10}
          x={0}
          y={55}
        />
        <rect
          fill="red"
          height="10"
          width={10}
          x={0}
          y={66}
        />
        <rect
          fill="red"
          height="10"
          width={10}
          x={0}
          y={77}
        />
        <rect
          fill="red"
          height="10"
          width={10}
          x={0}
          y={88}
        />
        <rect
          fill="red"
          height="10"
          width={10}
          x={0}
          y={99}
        />
        <rect
          fill="red"
          height="10"
          width={10}
          x={0}
          y={110}
        />
        <rect
          fill="red"
          height="10"
          width={10}
          x={0}
          y={121}
        />
        <rect
          fill="red"
          height="10"
          width={10}
          x={0}
          y={132}
        />
      </g>
    </svg>
  </div>
  <div
    className="content"
  >
    <div
      className="period"
    >
      <div>
        <div
          className="header"
        >
          <div>
            <div
              className="job"
            >
              <span>
                Senior
              </span>
              <span>
                Software Engineer
              </span>
            </div>
            <div
              className="companyName"
            >
              <div>
                test company name
              </div>
            </div>
          </div>
          <div
            className="dividerH"
          />
          <div
            style={
              Object {
                "fontSize": 24,
              }
            }
          >
            <div
              className="filled tag"
            >
              React
            </div>
          </div>
        </div>
      </div>
      <div
        className="time"
      >
        <div
          className="times"
        >
          <span>
            Jan, 2012
          </span>
          <span
            className="divider"
          />
          <span>
            Jan, 2013
          </span>
        </div>
      </div>
    </div>
    <div />
    <div>
      <ul />
    </div>
    <div>
      <ul
        className="achievements"
      />
    </div>
    <div
      className="projects"
    />
  </div>
</div>
`;
